### PR TITLE
Implement rclone-backed repository storage

### PR DESCRIPTION
**Pull request by karanhudia**
## Description
Adds rclone-backed repository storage across the backend and frontend. Borg operates against a server-derived local repository cache, and rclone mirrors that cache to the selected remote after successful Borg transactions or through explicit sync/hydrate actions.

Backend changes include rclone remote/storage/sync-job models, migration support, rclone command execution with redaction, availability and remote-management APIs, repository create/import/status/sync/hydrate flows, post-backup sync policy handling, failure-state persistence, managed config cleanup, safe path/flag validation, and PostgreSQL-compatible migration DDL. Managed rclone remotes write sections into the server-owned `RCLONE_CONFIG_ROOT/rclone.conf`, and the Cloud Storage API now supports managed remote edit/delete with config rollback and in-use delete protection.

Frontend changes add a left-sidebar `Cloud Storage` page under BACKUP for reusable rclone remotes. The page supports add, test, browse, edit, delete, rclone availability, empty/unavailable states, and repository usage counts using card treatment aligned with repository cards. Repository setup keeps the existing Location step and exposes `Cloud Storage` as a fourth location card beside Borg UI Server, Remote Client, and Managed Agent; selecting it shows the repository-owned rclone fields below the cards.

Docs reflect the Cloud Storage information architecture, repository-owned sync metadata, and the explicit out-of-scope boundary for enabling cloud mirrors on existing local repositories.

## Related Issue
Linear: https://linear.app/nullcodeai/issue/BOR-66/implement-rclone-backed-repository-storage

## Type of Change
- [x] 🐛 Bug fix (non-breaking change which fixes an issue)
- [x] ✨ New feature (non-breaking change which adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] 📖 Documentation update
- [x] 🎨 UI/UX improvement
- [x] 🧪 Test addition/improvement
- [x] ♻️ Refactoring

## Testing
- [x] I have tested this locally
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] All existing tests pass

Latest Code Review Reply validation:
- `git diff --check`
- `ruff check app tests`
- `ruff format --check app tests`
- `pytest tests/unit/test_api_rclone.py -q`
- `cd frontend && npm run check:locales`
- `cd frontend && npm run format:check`
- `cd frontend && NODE_OPTIONS=--max-old-space-size=4096 npm run typecheck`
- `cd frontend && NODE_OPTIONS=--max-old-space-size=4096 npm run lint`
- `cd frontend && NODE_OPTIONS=--max-old-space-size=3072 ROLLUP_MAX_PARALLEL_FILE_OPS=1 npm run build`
- `cd frontend && NODE_OPTIONS=--max-old-space-size=4096 npx vitest run src/pages/__tests__/CloudStorage.test.tsx --maxWorkers=1 --no-file-parallelism --reporter=verbose`
- `cd frontend && NODE_OPTIONS=--max-old-space-size=3072 ROLLUP_MAX_PARALLEL_FILE_OPS=1 npm run snapshots`
- Docker dev backend walkthrough on `127.0.0.1:8097` with a local rclone alias remote: rclone availability, managed remote create/test/browse/edit/delete, rclone-backed repository create/status/sync/hydrate, used-remote delete protection, import, and cleanup all returned the expected connected/current state.

Earlier implementation and review-reply passes also validated the original backend rclone workflows, migration paths, backup-service post-sync behavior, repository wizard payload handling, unavailable-state behavior, add-remote dialog flow, and Storybook snapshot generation.

## Checklist
- [x] I have read the [CONTRIBUTING](CONTRIBUTING.md) guidelines
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] New and existing unit tests pass locally with my changes

## Screenshots (if applicable)
Storybook snapshots cover the Cloud Storage page `overview`, `empty`, `unavailable`, `editing-remote`, `delete-confirmation`, and `browse-dialog` states under `frontend/storybook-snapshots/`. Existing repository Location step rclone states remain covered there as well.

## Additional Context
Direct Borg-over-rclone-mount behavior remains intentionally out of scope. The shipped implementation follows the approved BOR-65 local-cache-plus-rclone-mirror architecture.

Existing local repositories do not gain an in-place “Enable cloud mirror” action in this PR because safe conversion needs migration, rollback, cache-path, and conflict-handling design. BOR-67 tracks that follow-up and is linked from BOR-66.

The latest local frontend build still emits the repository’s pre-existing Vite large-chunk warning and webauthn mixed static/dynamic import warning; no new build warning was introduced for this review response.

---

**Note:** By submitting this pull request, you agree that your contributions will be licensed under the GNU General Public License v3.0, the same license as this project.

### CODE REVIEW

**Comment by karanhudia** [comment]
1. Got error in migration-

```
INFO:app.database.migrations:{"event": "Running migration: 113_add_rclone_storage", "logger": "app.database.migrations", "level": "info", "timestamp": "2026-05-25T06:07:13.380144Z"}
ERROR:app.database.migrations:{"error": "'Connection' object has no attribute 'get_bind'", "event": "Migration failed: 113_add_rclone_storage", "logger": "app.database.migrations", "level": "error", "timestamp": "2026-05-25T06:07:13.380190Z"}
INFO:app.database.migrations:{"event": "All migrations completed", "logger": "app.database.migrations", "level": "info", "timestamp": "2026-05-25T06:07:13.380259Z"}
```

2. Seems pretty congested, plus hover effect is wrong because of height of card increasing.

image
Even the icon is looking so weird because of congested. Use ui-ux-pro-max skill for design recommendation here within the scope of same design language.

3. rclone in repositories is disabled.

File: tests/unit/test_api_rclone.py [hunk 1]
```diff
@@ -0,0 +1,553 @@
...
+    repository = (
+        test_db.query(Repository).filter(Repository.name == "Imported Cloud Repo").one()
+    )
+    storage = (
+        test_db.query(RepositoryStorage)
+        .filter(RepositoryStorage.repository_id == repository.id)
+        .one()
+    )
+    assert storage.cache_path == str(tmp_path / "cache" / "repositories" / "1")
+
```
[the remark this answers is not in the document]
Comment (karanhudia) [reply tests/unit/test_api_rclone.py:807]: [codex] Verified on current head c2ae9fa: this assertion now builds the expected cache path with `str(repository.id)` in `tests/unit/test_api_rclone.py`, so there is no remaining hard-coded repository id in this test.

**Comment by karanhudia** [comment]
[codex] Addressed the latest owner feedback in c2ae9fa: migration 113 now supports the SQLAlchemy `Connection` used by `run_migrations()`, app Docker builds now point at the rclone-enabled runtime base tag, and the rclone repository-location card uses a tighter stable hover/selected treatment with refreshed Storybook snapshots. Local validation and snapshot commands are listed in the PR body.

File: app/api/rclone.py [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+        if config_file is not None:
+            config_file.unlink(missing_ok=True)
+        if isinstance(exc, HTTPException):
+            raise
+        raise HTTPException(
+            status_code=500,
+            detail={
+                "key": "backend.errors.rclone.failedToCreateRemote",
+                "message": str(exc) or exc.__class__.__name__,
+            },
```
[the remark this answers is not in the document]
Comment (karanhudia) [reply app/api/rclone.py]: [codex] I will keep the internal exception chained for logs but return a stable generic failure detail to clients instead of exposing str(exc).

File: app/api/rclone.py [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+    existing = db.query(RcloneRemote).filter(RcloneRemote.name == remote_name).first()
+    if existing:
+        raise HTTPException(
+            status_code=409, detail={"key": "backend.errors.rclone.remoteExists"}
+        )
+
+    provider = payload.provider.strip()
+    remote = RcloneRemote(
+        name=remote_name,
+        provider=provider,
```
[the remark this answers is not in the document]
Comment (karanhudia) [reply app/api/rclone.py]: [codex] I will add provider normalization/validation before the remote row is created so whitespace-only providers return a 400 with a stable rclone error key.

File: app/api/rclone.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (karanhudia) [reply app/api/rclone.py]: [codex] Fixed in 8f21d46: create-remote unexpected failures now return only `backend.errors.rclone.failedToCreateRemote` while preserving the chained exception internally. Updated the commit-failure test to assert the sanitized response; targeted backend validation passes.

File: app/api/rclone.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (karanhudia) [reply app/api/rclone.py]: [codex] Fixed in 8f21d46: provider is now normalized through `_normalize_provider()` and whitespace-only values return `400` with `backend.errors.rclone.invalidProvider`. Added `test_create_rclone_remote_rejects_blank_provider`; targeted backend validation passes.

**Comment by karanhudia** [comment]
How do I add rclone remotes?

**Comment by karanhudia** [comment]
[codex] Added the rclone remote creation path in f328eab4797d8c3be3963569297beac371f75824. In repository setup, choose Cloud storage, then use the new "Add remote" button beside the rclone remote selector. The dialog creates a managed remote via the existing backend API, validates the name/provider/config JSON, inserts the new remote into the selector, and selects it for the current setup flow. Storybook now covers the empty add-remote state and the create/error dialog states.

**Comment by karanhudia** [comment]
Why not have separate tab for rclone backend destinations i guess? Something with a friendly name, maybe cloud sources, etc. Idk you can be liberal and creative here according to convention.

**Comment by karanhudia** [comment]
image
Its clear that dropdown height is less here. Please make it as big as the button right to it.

**Comment by karanhudia** [comment]
[codex] Addressed the latest UI feedback in 9ae613a760c7b0127ac3a4d6fb16336737f89eb9. Repository setup now separates filesystem/SSH targets from rclone under a dedicated "Cloud sources" tab, and the rclone remote selector is height-aligned with the adjacent Add remote button. Updated focused wizard tests, locale labels, Storybook story coverage, and snapshots; local frontend validation is recorded in the PR body.

**Comment by karanhudia** [comment]
I meant that the tab for cloud sources should be in left sidebar where repositories are also there. The idea is simple, just like we have remote machines, agents, repositories in left sidebar. cloud backup sources are also a big desitnation and deserve a tab where we can use the same card like repositories, with same design language (use ui ux pro max skill). Then we wont need to pollute the create backup plan wizard with all this stuff. It would be simple choice for user then.

Consider these. Some might be irrelavant and you can ignore them, but if something makes sense then do it. Update the engineering docs accordingly.
  - Add a new sidebar tab under BACKUP called Cloud Storage.
  - Cloud Storage manages reusable rclone remotes only.
  - In Cloud Storage, user can add remote, choose provider, enter config, test connection, browse remote, and see
    usage count.
  - Do not put repository creation in Cloud Storage.
  - In Repository Wizard, do not add a new step.
  - Use the existing Location step.
  - Add Cloud Storage as a fourth location card beside:
    Borg UI Server, SSH Server, Managed Agent.
  - When Cloud Storage is selected, show rclone fields below the location cards:
    Rclone remote, Relative remote path, Local cache path readonly, Sync policy, Advanced flags.
  - For new cloud backups, user creates a repository with location = Cloud Storage.
  - For existing local repositories, add action on repository card/settings: Enable cloud mirror.
  - Enabling cloud mirror should ask for rclone remote, relative remote path, sync policy, optional flags.
  - Do not call this “add sync policy”; call it Enable cloud mirror.
  - Backup Plan Wizard should not get a cloud step.
  - In Backup Plan Wizard’s existing Repositories step, only show metadata chips:
    Cloud Storage, r2-prod:borg-ui/photos, Synced/Pending/Failed.
  - Backup Plans only select repositories. Repositories handle cloud sync.
  - Backend model:
    rclone_remotes stores provider/config/status.
    repository_storage stores repo cloud settings.
  - Repository owns:
    rclone_remote_id, rclone_remote_path, cache_path, sync_policy, sync_status.
  - Runtime:
    Borg writes to local cache path.
    After successful Borg operation, run rclone sync to remote.
    If cache is missing/stale before restore/browse/check, run hydrate from remote first.
  - MVP should not use direct rclone mount.
  - MVP should not use Borg 2 native rclone:remote:path.
  - Those can be advanced later features.

**Comment by karanhudia** [comment]
[codex] Addressed the latest Cloud Storage IA feedback in commit 72a43ad.

- Added a left-sidebar BACKUP > Cloud Storage page for reusable rclone remotes only: add remote, provider/config metadata, test connection, browse remote, rclone availability, and repository usage count.
- Kept repository creation in the existing Repository Wizard Location step and changed it to four location cards: Borg UI Server, SSH Server, Managed Agent, and Cloud Storage. Rclone fields now render below the cards when Cloud Storage is selected.
- Left the Backup Plan Wizard unchanged; cloud sync remains repository-owned metadata/status.
- Updated the engineering spec/plan to reflect this IA and created BOR-67 for existing-local “Enable cloud mirror” conversion because safe migration/rollback/cache-path handling is a separate design from BOR-66.
- Fixed managed rclone config wiring so created remotes are written into the server-owned /data/rclone/rclone.conf used by runtime rclone commands.

Validation for the pushed commit includes backend ruff/pytest, frontend locales/format/typecheck/lint/build/Vitest, Storybook snapshots, and a disposable app walkthrough using a local rclone alias remote for create, test, browse, repository create, status, sync, hydrate, import, and usage count.

**Comment by karanhudia** [comment]
Looks fine now. Just minor UI feedback before we merge.

image

Please consult ui ux pro max skill to update this page. This does not match the design language at all. The top header is useless.  the cards should be just like repository cards. Probably use the same without duplication.

There should be edit and delete option for them as well.

**Comment by karanhudia** [comment]
[codex] Addressed the Cloud Storage review feedback in bb5ae5f. The page now drops the redundant top header, uses a shared `OperationalCard` shell also adopted by `RepositoryCard`, and renders remotes as repository-card-style cards. Remote cards now include test, browse, edit, and delete controls, with delete disabled for remotes used by repositories; backend PUT/DELETE endpoints keep managed rclone config sections in sync and block in-use deletes. Storybook snapshots cover overview, empty, unavailable, edit, delete confirmation, and browse states; local validation and the live rclone walkthrough are listed in the PR body.
